{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "# -*- coding: utf-8 -*-\n",
    "\"\"\"\n",
    "Created on Fri Dec 20 20:10:22 2019\n",
    "\n",
    "@author: raoulrosenthal\n",
    "\"\"\"\n",
    "\n",
    "# -*- coding: utf-8 -*-\n",
    "\"\"\"Amino acid mutation.ipynb\n",
    "\n",
    "Automatically generated by Colaboratory.\n",
    "\n",
    "Original file is located at\n",
    "    https://colab.research.google.com/drive/1dQGGuQUbD7YdX1oVSAoi_lOYuJcaKueS\n",
    "\"\"\"\n",
    "\n",
    "# Run this part to upload files from local\n",
    "\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio import Align\n",
    "from Bio.Align import AlignInfo\n",
    "from Bio import AlignIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio import pairwise2\n",
    "from Bio.Alphabet import generic_dna\n",
    "import glob\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def string_replace(string, new_character, position):\n",
    "    return(string[0:position] + str(new_character) + string[position+1:])\n",
    "\n",
    "def import_sequences(file_path, type_='fastq'): #imports sequences from file \n",
    "    Tracefiles = glob.glob(file_path)\n",
    "    list_of_sequences=[]\n",
    "    for file in Tracefiles:\n",
    "        for sequence in SeqIO.parse(file, type_):\n",
    "            list_of_sequences.append(sequence)\n",
    "    return(list_of_sequences)\n",
    "\n",
    "def reverse_sequences(list_of_sequences, string_to_reverse): #a function that returns the reverse complement of sequences that have a given string in their file name\n",
    "    for i in range(len(list_of_sequences)):\n",
    "        sequence = list_of_sequences[i].seq\n",
    "        if(len(sequence)) > 25:\n",
    "            if string_to_reverse in list_of_sequences[i].id:            # find all reverse traces\n",
    "                list_of_sequences[i].seq = sequence.reverse_complement()\n",
    "    return(list_of_sequences)\n",
    "\n",
    "def load_reference(file_path, type_ = \"genbank\", sequence_number = 0): #loads reference sequence\n",
    "    reference_plasmid = [i for i in SeqIO.parse(file_path, type_)][sequence_number]\n",
    "    reference_plasmid = reference_plasmid\n",
    "    return(reference_plasmid.seq)\n",
    "\n",
    "def find_SNPs(list_of_sequences, reference_sequence, alignment_threshold = -150): # finds all the SNPs in a given variant (both forward and reverse sequence) in the overlap region the SNP with a higher phred score is taken the list of sequences \n",
    "    list_of_SNPS = [] \n",
    "    not_aligned = 0 \n",
    "    for i in range(0,len(list_of_sequences),2):\n",
    "#        counter += 1\n",
    "        read_1 = list_of_sequences[i].seq[:] # read in forward sequence\n",
    "        read_2 = list_of_sequences[i+1].seq[:] # read in reverse sequence\n",
    "        try: # align both reads to reference sequence \n",
    "            alignment1 = pairwise2.align.globalms(reference_gene,read_1,2,-0.5,-4,-2, one_alignment_only =1)[0]\n",
    "            alignment2 = pairwise2.align.globalms(reference_gene,read_2,2,-0.5,-4,-2, one_alignment_only =1)[0]\n",
    "\n",
    "        except IndexError:\n",
    "            not_aligned += 1\n",
    "            print('sequence', i, ' is not well aliged')\n",
    "            continue\n",
    "        mutation_position  = []\n",
    "        reference_base = []\n",
    "        mutated_base = [] \n",
    "        if alignment1[-3]<alignment_threshold or alignment2[-3]<alignment_threshold: # check for good quality alignments i.e. score > 50\n",
    "            not_aligned += 1\n",
    "            print('sequence', i, ' is not well aliged')\n",
    "        else:\n",
    "            forward_seq = alignment1[1]\n",
    "            reverse_seq = alignment2[1]\n",
    "            assembeled_gene = str()\n",
    "    #        print(len(forward_seq), len(reverse_seq))\n",
    "            for nucleotide in range(len(reference_gene)): # index bases according to reference sequence\n",
    "                if (forward_seq[nucleotide] != '-' \n",
    "                    and forward_seq[nucleotide] != reference_gene[nucleotide] \n",
    "                    and reverse_seq[nucleotide] == '-'): # only compares mutations where there is a single forward read\n",
    "\n",
    "                    mutation_position.append(nucleotide)\n",
    "                    reference_base.append(reference_gene[nucleotide])\n",
    "                    mutated_base.append(forward_seq[nucleotide])\n",
    "\n",
    "                elif (reverse_seq[nucleotide] != '-' \n",
    "                      and reverse_seq[nucleotide] != reference_gene[nucleotide] \n",
    "                      and forward_seq[nucleotide] == '-'): # only compares mutations where there is a single reverse read\n",
    "\n",
    "                    mutation_position.append(nucleotide)\n",
    "                    reference_base.append(reference_gene[nucleotide])\n",
    "                    mutated_base.append(reverse_seq[nucleotide])\n",
    "\n",
    "                elif (reverse_seq[nucleotide] != '-' and forward_seq[nucleotide] != '-'): # compares mutations where there is double coverage\n",
    "\n",
    "\n",
    "                    if (reverse_seq[nucleotide] != reference_gene[nucleotide] \n",
    "                        or forward_seq[nucleotide] != reference_gene[nucleotide]): # if only one read is mismatched with the ref, take strand with better phred score \n",
    "\n",
    "                        aligned_forward_seq = forward_seq[0:nucleotide].strip('-')\n",
    "                        aligned_reverse_seq = reverse_seq[0:nucleotide].strip('-')\n",
    "                        forward_position = len(aligned_forward_seq)\n",
    "                        reverse_position = len(aligned_reverse_seq)\n",
    "                        phred_score_F = recordlist[i].letter_annotations['phred_quality'][forward_position-aligned_forward_seq.count('-')]\n",
    "                        phred_score_R = recordlist[i+1].letter_annotations['phred_quality'][reverse_position-aligned_reverse_seq.count('-')]\n",
    "\n",
    "                        if phred_score_F >= phred_score_R:\n",
    "                            if forward_seq[nucleotide] != reference_gene[nucleotide]:\n",
    "                                mutation_position.append(nucleotide)\n",
    "                                reference_base.append(reference_gene[nucleotide])\n",
    "                                mutated_base.append(forward_seq[nucleotide])\n",
    "                            else:\n",
    "                                continue\n",
    "                        if phred_score_F < phred_score_R:\n",
    "                            if reverse_seq[nucleotide] != reference_gene[nucleotide]:\n",
    "                                mutation_position.append(nucleotide)\n",
    "                                reference_base.append(reference_gene[nucleotide])\n",
    "                                mutated_base.append(reverse_seq[nucleotide])\n",
    "                            else: \n",
    "                                continue\n",
    "                    elif (reverse_seq[nucleotide] != reference_gene[nucleotide] \n",
    "                          and forward_seq[nucleotide] != reference_gene[nucleotide]): # if strands are mismatched with the reference, it does not matter with strand is taken to annotate the mutation \n",
    "\n",
    "                        if reverse_seq[nucleotide] == forward_seq[nucleotide]:\n",
    "                            mutation_position.append(nucleotide)\n",
    "                            reference_base.append(reference_gene[nucleotide])\n",
    "                            mutated_base.append(forward_seq[nucleotide])\n",
    "\n",
    "            list_of_SNPS.append([mutation_position, reference_base, mutated_base])\n",
    "    \n",
    "    print('\\n', not_aligned,'out of', len(list_of_sequences)//2, ' sequences could not be aligned well')\n",
    "    return(list_of_SNPS)\n",
    "\n",
    "def classify_mutations(mutations): # counts mutsations by transition type\n",
    "    AC, AG, AT, CA, CG, CT, GA, GC, GT, TA, TC, TG = [0]*12 \n",
    "    all_A = 0\n",
    "    all_C = 0\n",
    "    all_G = 0\n",
    "    all_T = 0\n",
    "    all_GC = 0\n",
    "    all_AT = 0\n",
    "    counter = 0\n",
    "    for i in mutations:\n",
    "        for j in range(len(i[1])):\n",
    "            r = i[1][j]\n",
    "            m = i[2][j]\n",
    "            if r == 'A' and m =='C':\n",
    "                AC += 1\n",
    "            if r == 'A' and m =='G':\n",
    "                AG += 1\n",
    "            if r == 'A' and m =='T':\n",
    "                AT += 1\n",
    "            if r == 'C' and m =='A':\n",
    "                CA += 1\n",
    "            if r == 'C' and m =='G':\n",
    "                CG += 1\n",
    "            if r == 'C' and m =='T':\n",
    "                CT += 1\n",
    "            if r == 'G' and m =='A':\n",
    "                GA += 1\n",
    "            if r == 'G' and m =='C':\n",
    "                GC += 1\n",
    "            if r == 'G' and m =='T':\n",
    "                GT += 1\n",
    "            if r == 'T' and m =='A':\n",
    "                TA += 1\n",
    "            if r == 'T' and m =='C':\n",
    "                TC += 1\n",
    "            if r == 'T' and m =='G':\n",
    "                TG += 1\n",
    "            counter +=1\n",
    "    all_A = AC+AG+AT\n",
    "    all_C = CA+CG+CT\n",
    "    all_G = GA+GC+GT\n",
    "    all_T = TA+TC+TG\n",
    "    all_GC = all_C + all_G\n",
    "    all_AT = all_A + all_T\n",
    "    print('total # mutations', counter, '\\n',AC, AG, AT, CA, CG, CT, GA, GC, GT, TA, TC, TG, '\\n', \n",
    "          'total A:', all_A, (all_A)/ counter, '\\n', \n",
    "          'total C:', all_C, (all_C)/ counter, '\\n',\n",
    "          'total G:', all_G, (all_G)/ counter, '\\n', \n",
    "          'total T:', all_T, (all_T)/ counter, '\\n',\n",
    "          'total GC:', all_GC, all_GC/ counter, '\\n',\n",
    "          'total AT:', all_AT, all_AT/ counter, '\\n',\n",
    "         )\n",
    "    return(all_AT, all_GC, all_A, all_C, all_G, all_T)\n",
    "\n",
    "def plot_mutation_distribution(list_of_SNPS): # plots a histogram of the number of mutations per sequence\n",
    "    mutation_number = []\n",
    "    for SNP in list_of_SNPS:\n",
    "        mutation_number.append(len(SNP[0]))\n",
    "    fig, ax = plt.subplots(1,1, figsize=(6,6))\n",
    "    ax.hist(mutation_number, bins=max(mutation_number)*2+1, density = False)\n",
    "    ax.set(title='mutation distribution ', xlabel= '# of mutations per sequence', ylabel='# of sequences')\n",
    "    return(plt.show())\n",
    "\n",
    "def count_amino_acid_transitions(amino_acid_mutations, amino_acid): # counts the number of mutations that transitioned the sequence into a given amino acid\n",
    "    count_amino_acid = 0 \n",
    "    for mutation in amino_acid_mutations:\n",
    "        count_amino_acid += mutation[-1].count(str(amino_acid))\n",
    "    return(count_amino_acid)\n",
    "\n",
    "def find_AA_mutations_from_SNPs(list_of_SNPS, reference_gene): # translates reference and find amino acid mutations that correspond to SNPs\n",
    "    if type(reference_gene) != type(str()):\n",
    "        return(print('reference_gene must be a string'))\n",
    "    else:\n",
    "        reference_gene = Seq(str(reference_gene), generic_dna) \n",
    "        reference_gene_T = str()\n",
    "        mutations_AA = [] \n",
    "        for SNPS_in_sequence in list_of_SNPS:\n",
    "            original_AA = []\n",
    "            mutated_AA = []\n",
    "            for SNP in range(len(SNPS_in_sequence[0])):\n",
    "                original_AA.append(reference_gene.translate()[(SNPS_in_sequence[0][SNP])//3])\n",
    "                reference_gene = string_replace(reference_gene, SNPS_in_sequence[2][SNP], SNPS_in_sequence[0][SNP])\n",
    "                mutated_AA.append(reference_gene.translate()[(SNPS_in_sequence[0][SNP])//3])\n",
    "            mutations_AA.append([original_AA, mutated_AA])\n",
    "        return(mutations_AA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sequence 6  is not well aliged\n",
      "sequence 28  is not well aliged\n",
      "sequence 30  is not well aliged\n",
      "sequence 58  is not well aliged\n",
      "sequence 62  is not well aliged\n",
      "\n",
      " 5 out of 48  sequences could not be aligned well\n"
     ]
    }
   ],
   "source": [
    "file_path = \"/Users/raoulrosenthal/Downloads/96 documents from New Folder.fastq\"\n",
    "file_path_ref = \"/Users/raoulrosenthal/Documents/Geneious 11.0 Data/temp/116/p15A_trmI_AldO.gb\"\n",
    "list_of_sequences = import_sequences(file_path)\n",
    "list_of_sequences = reverse_sequences(list_of_sequences, 'BAD2')\n",
    "reference_plasmid = load_reference(file_path_ref)\n",
    "reference_gene = str(reference_plasmid)[3643:4900]\n",
    "list_of_SNPS = find_SNPs(list_of_sequences, reference_gene, alignment_threshold = -150)\n",
    "del(list_of_SNPS[15])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total # mutations 70 \n",
      " 0 2 1 7 2 19 21 2 4 4 8 0 \n",
      " total A: 3 0.04285714285714286 \n",
      " total C: 28 0.4 \n",
      " total G: 27 0.38571428571428573 \n",
      " total T: 12 0.17142857142857143 \n",
      " total GC: 55 0.7857142857142857 \n",
      " total AT: 15 0.21428571428571427 \n",
      "\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYIAAAGDCAYAAAAmphcsAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvhp/UCwAAHFBJREFUeJzt3Xu4bXVd7/H3R0CuIiLLC8J2e0nMehRt60lRI0VFwEtFFAdMvLS7mOgpIzRTMzMe66iZ1Tk7BTxhKHlLRAQiyUhDNvebqOFOEGSDoAIqgn7PH2OsnCzXZe7LHHPv9Xu/nmc+a44xf/P3+8659p6fNS7zN1JVSJLada9pFyBJmi6DQJIaZxBIUuMMAklqnEEgSY0zCCSpcQaBlpUkr0vyngHGWZmkkmzbL5+e5MWbqe+nJbl6ZHldkgM2R999f1ck2X9z9aetX/wegbYUSd4EPLKqjhyz/f7ASVW11yTrWmDslcBXgO2q6u4NeF4BP1FVX96A56wDXl5V/7yBZZLkROC6qnr9hj5X7XCLQNqCzG5hSEMyCLRJ+t0Wv5/k0iR3JHlvkgf2u0puS/LPSe7Xt90/yXXzPP+AJAcCrwN+JcntSS7pH39Jkqv6vq5J8hv9+p2B04E9+/a3J9kzyZuSnDTS//P7XSHfTHJOkp+cM/Zr+tq/leSDSXZY4HVuk+Qvktyc5Brg4DmPn5Pk5f39Ryb5177Pm5N8sF//mb75JX29vzL7niT5gyRfB06Y730CnpjkyiS3Jjlhts4kRyU5d04t1dewGjgCOKYf79TR97y/v32Sdya5vr+9M8n2o7+vJL+XZH2SG5K8ZOF/DdpaGQTaHH4JeBbwKOB5dB/QrwP2oPs3dvRSHVTVp4C3Ah+sql2q6nH9Q+uBQ4BdgZcA70jyhKq6A3gucH3ffpequn60zySPAk4GXg3MAJ8ETk1y75FmhwEHAg8DHgsctUCJv97X8XhgFXDoIi/nT4AzgfsBewF/1b/Gp/ePP66v94P98oOA3YGHAqsX6PMI4DnAI+je5yV39VTVGuD9wNv68Z43T7M/BH4W2Bd4HPCkOX0/CLgv8BDgZcBfzwa7lg+DQJvDX1XVjVX1NeDfgPOq6qKquhP4KN2H50apqtOq6j+r8690H7BPG/PpvwKcVlVnVdVdwF8AOwJPGWnzrqq6vqpuAU6l+0Ccz2HAO6vq2r7tny0y7l10H+p7VtX3qurcRdoC/BB4Y1XdWVXfXaDNu0fG/lPg8CX6HNcRwJuran1V3QT8MfCikcfv6h+/q6o+CdwO7LOZxtYWwiDQ5nDjyP3vzrO8y8Z2nOS5Sf4jyS1JvgkcRLelMY49gf+aXaiqHwLX0v11O+vrI/e/s0ite/bPnfVfC7QDOAYI8Pl+t9RLl6jzpqr63hJt5o695xLtx3WP92ievr8x52D4Yu+RtlIGgYZ0B7DT7EKSbeh22cy6xyls/b7qD9P9Jf/AqtqNbvdO5ms/j+vp/jKf7S/A3sDXNqL2G/rnzlqxUMOq+npV/XpV7Qn8BvA3SR65SN/jnLo3d+zZ3WBz39MHbWDf93iP5vStRhgEGtIXgR2SHJxkO7p90duPPH4jsDLJ7L/Le/eP3wTcneS5wLPntL9/kvsuMN4pwMFJntmP93vAncBnN6L2U4Cjk+zV7yM/dqGGSX45yewprbfSfRj/YKTmh2/E+K/ox96d7vjL7PGFS4CfSrJvfwD5TXOet9R4JwOvTzKTZA/gDcBJi7TXMmQQaDBV9S3gt4H30P1VfgcwenbMP/Y/v5Hkwqq6je5A8yl0H6j/E/j4SH9foPsgu6Y/K+geu0uq6mrgSLqDtTfTHch+XlV9fyPK/zvgDLoP3guBjyzS9onAeUlu7+t9VVV9pX/sTcD7+noP24Dx/4Hu+Mg1/e0tAFX1ReDNwD8DXwLmHo94L/CYfryPzdPvW4C1wKXAZf1re8sG1KVlwC+USVLj3CKQpMYZBJLUOINAkhpnEEhS4wwCSWrcVjHT4R577FErV66cdhmStFW54IILbq6qmaXabRVBsHLlStauXTvtMiRpq5JksalQ/pu7hiSpcQaBJDXOIJCkxhkEktQ4g0CSGmcQSFLjDAJJapxBIEmNMwgkqXEGgSQ1bmJBkOT4JOuTXD7PY69JUv01UiVJUzTJLYITgQPnrkyyN/As4KsTHFuSNKaJBUFVfQa4ZZ6H3gEcA3ixZEnaAgw6+2iS5wNfq6pLkizVdjWwGmDFihUDVLd1WnnsaYONte64gwcbS9JwBjtYnGQn4A+BN4zTvqrWVNWqqlo1M7PkdNqSpI005FlDjwAeBlySZB2wF3BhkgcNWIMkaY7Bdg1V1WXAA2aX+zBYVVU3D1WDJOnHTfL00ZOBzwH7JLkuycsmNZYkaeNNbIugqg5f4vGVkxpbkjQ+v1ksSY0zCCSpcQaBJDXOIJCkxhkEktQ4g0CSGmcQSFLjDAJJapxBIEmNMwgkqXEGgSQ1ziCQpMYZBJLUOINAkhpnEEhS4wwCSWqcQSBJjTMIJKlxBoEkNc4gkKTGGQSS1DiDQJIaZxBIUuMMAklqnEEgSY0zCCSpcQaBJDXOIJCkxhkEktQ4g0CSGmcQSFLjDAJJapxBIEmNMwgkqXEGgSQ1bmJBkOT4JOuTXD6y7s+TfCHJpUk+mmS3SY0vSRrPJLcITgQOnLPuLOCnq+qxwBeB105wfEnSGCYWBFX1GeCWOevOrKq7+8X/APaa1PiSpPFM8xjBS4HTpzi+JAnYdhqDJvlD4G7g/Yu0WQ2sBlixYsVAlWlLsvLY0wYdb91xBw86nrSlGHyLIMmLgUOAI6qqFmpXVWuqalVVrZqZmRmuQElqzKBbBEkOBP4A+Lmq+s6QY0uS5jfJ00dPBj4H7JPkuiQvA94N3Ac4K8nFSf7PpMaXJI1nYlsEVXX4PKvfO6nxJEkbx28WS1LjDAJJapxBIEmNMwgkqXEGgSQ1ziCQpMYZBJLUOINAkhpnEEhS4wwCSWqcQSBJjTMIJKlxBoEkNc4gkKTGGQSS1DiDQJIaZxBIUuMMAklqnEEgSY0zCCSpcQaBJDXOIJCkxhkEktQ4g0CSGmcQSFLjDAJJapxBIEmNMwgkqXEGgSQ1ziCQpMYZBJLUOINAkhpnEEhS4wwCSWqcQSBJjZtYECQ5Psn6JJePrNs9yVlJvtT/vN+kxpckjWeSWwQnAgfOWXcscHZV/QRwdr8sSZqiiQVBVX0GuGXO6hcA7+vvvw944aTGlySNZ+hjBA+sqhsA+p8PGHh8SdIcW+zB4iSrk6xNsvamm26adjmStGwNHQQ3JnkwQP9z/UINq2pNVa2qqlUzMzODFShJrRk6CD4OvLi//2LgnwYeX5I0xyRPHz0Z+BywT5LrkrwMOA54VpIvAc/qlyVJU7TtpDquqsMXeOiZkxpTkrThttiDxZKkYRgEktQ4g0CSGmcQSFLjDAJJapxBIEmNMwgkqXEGgSQ1ziCQpMYZBJLUOINAkhpnEEhS4wwCSWqcQSBJjTMIJKlxBoEkNW6DgiDJ/ZI8dlLFSJKGt2QQJDknya5JdgcuAU5I8vbJlyZJGsI4WwT3rapvA78InFBVPwMcMNmyJElDGScItk3yYOAw4BMTrkeSNLBxguDNwBnAf1bV+UkeDnxpsmVJkoay7VINquofgX8cWb4G+KVJFiVJGs44B4sfleTsJJf3y49N8vrJlyZJGsI4u4b+DngtcBdAVV0K/Ooki5IkDWecINipqj4/Z93dkyhGkjS8cYLg5iSPAAogyaHADROtSpI0mCUPFgOvANYAj07yNeArwJETrUqSNJhxzhq6Bjggyc7AvarqtsmXJUkayjhnDb01yW5VdUdV3dbPN/SWIYqTJE3eOMcInltV35xdqKpbgYMmV5IkaUjjBME2SbafXUiyI7D9Iu0lSVuRcQ4WnwScneQEujOHXgq8b6JVSZIGM87B4rcluQx4JhDgT6rqjIlXJkkaxDhbBFTV6cDpE65FkjQF45w19ItJvpTkW0m+neS2JN8eojhJ0uSNs0XwNuB5VXXVpIuRJA1vnLOGbtzcIZDkfyW5IsnlSU5OssPm7F+SNL5xtgjWJvkg8DHgztmVVfWRjRkwyUOAo4HHVNV3k5xCN5vpiRvTnyRp04wTBLsC3wGePbKugI0KgpFxd0xyF7ATcP0m9CVJ2gTjnD76ks05YFV9LclfAF8FvgucWVVnzm2XZDWwGmDFihWbswRJ0ojBr1CW5H7AC4CHAXsCOyf5sdlMq2pNVa2qqlUzMzMbO5wkaQnTuELZAcBXquqmqrqLbhfTUzahP0nSJpjGFcq+Cvxskp2ShO4by56aKklTMvgVyqrqPOBDwIXAZX0Naza2P0nSppnKFcqq6o3AGzelD0nS5uEVyiSpcUsGQZI3zFkGoKrePKGaJEkDGmfX0B0j93cADsGDu5K0bIyza+h/jy73Xwb7+MQqkiQNapyzhubaCXj45i5EkjQd4xwjuIz+1FFgG2AG8PiAJC0T4xwjOGTk/t1001JvyhfKJElbkHGCYO7porvOnjkEUFW3bNaKJEmDGicILgT2Bm6lu3j9bnTTREC3y8jjBZK0FRvnYPGn6C5VuUdV3Z9uV9FHquphVWUISNJWbpwgeGJVfXJ2oapOB35uciVJkoY0zq6hm/vrD5xEtyvoSOAbE61KkjSYcbYIDqc7ZfSj/W2mXydJWgbG+WbxLcCrkuxSVbcPUJMkaUDjXKryKUmuBK7slx+X5G8mXpkkaRDj7Bp6B/Ac+uMCVXUJ8PRJFiVJGs44B4upqmtHv0QG/GAy5Wx+K489bdDx1h138KDjSdKmGicIrk3yFKCS3Bs4GqehlqRlY5xdQ79Jd7nKhwDXAfv2y5KkZWCcs4ZuBo4YoBZJ0hSMc9bQ25LsmmS7JGcnuTnJJl28XpK05Rhn19Czq+rbdHMMXQc8Cvj9iVYlSRrMOEGwXf/zIOBkp52WpOVlnLOGTk3yBeC7wG8nmQG+N9myJElDWXKLoKqOBZ4MrKqqu4DvAC+YdGGSpGGM+4WyW0fu3wHcMbGKJEmDGucYgSRpGVswCJLs1//cfrhyJElDW2yL4F39z88NUYgkaToWO0ZwV5ITgIckedfcB6vq6MmVJUkaymJBcAhwAPAM4IJhypEkDW3BIOjnGPpAkqv6axBIkpahcc4a+kaSjyZZn+TGJB9OstfEK5MkDWKcIDgB+DiwJ91U1Kf26yRJy8A4QfCAqjqhqu7ubycCMxOuS5I0kHGC4KYkRybZpr8dSX/94o2VZLckH0ryhSRXJXnypvQnSdp44wTBS4HDgK8DNwCH9us2xV8Cn6qqRwOPw0tfStLUjHOFsq8Cz99cAybZFXg6cFTf//eB72+u/iVJG2Yacw09HLgJOCHJRUnek2TnKdQhSWLM2UcnMOYTgFdW1XlJ/hI4Fvij0UZJVgOrAVasWDF4kdIkrTz2tEHHW3fcwYOOp63LNLYIrgOuq6rz+uUP0QXDPVTVmqpaVVWrZmY8SUmSJmWci9e/fuT+Js9EWlVfB65Nsk+/6pnAlZvaryRp4yw2DfUx/Wmdh46s3lwzkb4SeH+SS4F9gbdupn4lSRtosWMEVwO/DDw8yb/RneJ5/yT7VNXVmzJoVV0MrNqUPiRJm8diu4ZuBV4HfBnYnx9dn+DYJJ+dcF2SpIEstkVwIPBG4BHA24FLgDuq6iVDFCZJGsaCWwRV9bqqeiawDjiJLjRmkpyb5NSB6pMkTdg43yM4o6rOB85P8ltV9dQke0y6MEnSMJY8fbSqjhlZPKpfd/OkCpIkDWuDvlDmlcokafmZxjeLJUlbEINAkhpnEEhS4wwCSWqcQSBJjTMIJKlxBoEkNc4gkKTGGQSS1DiDQJIaZxBIUuMMAklqnEEgSY0zCCSpcQaBJDXOIJCkxhkEktQ4g0CSGmcQSFLjDAJJapxBIEmNMwgkqXEGgSQ1ziCQpMYZBJLUOINAkhpnEEhS4wwCSWqcQSBJjTMIJKlxUwuCJNskuSjJJ6ZVgyRpulsErwKumuL4kiSmFARJ9gIOBt4zjfElST8yrS2CdwLHAD9cqEGS1UnWJll70003DVeZJDVm8CBIcgiwvqouWKxdVa2pqlVVtWpmZmag6iSpPdPYItgPeH6SdcAHgGckOWkKdUiSmEIQVNVrq2qvqloJ/CrwL1V15NB1SJI6fo9Akhq37TQHr6pzgHOmWYMktc4tAklqnEEgSY0zCCSpcQaBJDXOIJCkxhkEktQ4g0CSGmcQSFLjDAJJapxBIEmNMwgkqXEGgSQ1ziCQpMYZBJLUOINAkhpnEEhS4wwCSWqcQSBJjTMIJKlxBoEkNc4gkKTGGQSS1DiDQJIaZxBIUuMMAklqnEEgSY0zCCSpcQaBJDXOIJCkxhkEktQ4g0CSGmcQSFLjDAJJapxBIEmNMwgkqXGDB0GSvZN8OslVSa5I8qqha5Ak/ci2UxjzbuD3qurCJPcBLkhyVlVdOYVaJKl5g28RVNUNVXVhf/824CrgIUPXIUnqTGOL4L8lWQk8HjhvnsdWA6sBVqxYMWhdkrSQlceeNuh46447eOJjTO1gcZJdgA8Dr66qb899vKrWVNWqqlo1MzMzfIGS1IipBEGS7ehC4P1V9ZFp1CBJ6kzjrKEA7wWuqqq3Dz2+JOmeprFFsB/wIuAZSS7ubwdNoQ5JElM4WFxV5wIZelxJ0vz8ZrEkNc4gkKTGGQSS1DiDQJIaZxBIUuMMAklqnEEgSY0zCCSpcQaBJDXOIJCkxhkEktQ4g0CSGmcQSFLjDAJJapxBIEmNMwgkqXEGgSQ1bvArlEla/lYee9pgY6077uDBxlqu3CKQpMYZBJLUOINAkhpnEEhS4wwCSWqcQSBJjTMIJKlxBoEkNc4gkKTGGQSS1DiDQJIaZxBIUuMMAklqnEEgSY0zCCSpcQaBJDXOIJCkxk0lCJIcmOTqJF9Ocuw0apAkdQYPgiTbAH8NPBd4DHB4kscMXYckqTONLYInAV+uqmuq6vvAB4AXTKEOSRLTCYKHANeOLF/Xr5MkTUGqatgBk18GnlNVL++XXwQ8qapeOafdamB1v7gPcPVGDrkHcPNGPndr5Wtug6+5DZvymh9aVTNLNdp2IzvfFNcBe48s7wVcP7dRVa0B1mzqYEnWVtWqTe1na+JrboOvuQ1DvOZp7Bo6H/iJJA9Lcm/gV4GPT6EOSRJT2CKoqruT/A5wBrANcHxVXTF0HZKkzjR2DVFVnwQ+OdBwm7x7aSvka26Dr7kNE3/Ngx8sliRtWZxiQpIat6yDoLWpLJIcn2R9ksunXcsQkuyd5NNJrkpyRZJXTbumSUuyQ5LPJ7mkf81/PO2ahpJkmyQXJfnEtGsZQpJ1SS5LcnGStRMda7nuGuqnsvgi8Cy6U1bPBw6vqiunWtgEJXk6cDvw/6rqp6ddz6QleTDw4Kq6MMl9gAuAFy7z33GAnavq9iTbAecCr6qq/5hyaROX5HeBVcCuVXXItOuZtCTrgFVVNfHvTSznLYLmprKoqs8At0y7jqFU1Q1VdWF//zbgKpb5t9Src3u/uF1/W55/zY1IshdwMPCeadeyHC3nIHAqi4YkWQk8HjhvupVMXr+L5GJgPXBWVS371wy8EzgG+OG0CxlQAWcmuaCfaWFilnMQZJ51y/4vpxYl2QX4MPDqqvr2tOuZtKr6QVXtS/et/CclWda7AZMcAqyvqgumXcvA9quqJ9DN1PyKftfvRCznIBhrKgtt3fr95B8G3l9VH5l2PUOqqm8C5wAHTrmUSdsPeH6/z/wDwDOSnDTdkiavqq7vf64HPkq3u3silnMQOJXFMtcfOH0vcFVVvX3a9QwhyUyS3fr7OwIHAF+YblWTVVWvraq9qmol3f/jf6mqI6dc1kQl2bk/AYIkOwPPBiZ2NuCyDYKquhuYncriKuCU5T6VRZKTgc8B+yS5LsnLpl3ThO0HvIjuL8SL+9tB0y5qwh4MfDrJpXR/7JxVVU2cTtmYBwLnJrkE+DxwWlV9alKDLdvTRyVJ41m2WwSSpPEYBJLUOINAkhpnEEhS4wwCSWqcQaCxJfmzJPsneeGGzuban/9+Xj975NM2c12vTrLThrZL8snZc/KllhkE2hD/g24un58D/m0Dn/tM4AtV9fiq2tDnLuXVwJJBMLddVR3Ufzt3i9PPnisNwiDQkpL8ef8FpifSfWHt5cDfJnnDPG0fmuTsJJf2P1ck2Rd4G3BQ/6WvHec8Z12Styb5XJK1SZ6Q5Iwk/5nkN/s2+4/OQ5/k3UmOSnI0sCfdl6w+3T/2t30//z1f/wLt1iXZo7//u0ku72+v7tet7K918Hd9X2fO1p7k6CRX9q/zA/O8D0cl+ackn0p3TYw3jjx2ZH9NgYuT/N/ZD/0ktyd5c5LzgCfP6e/Hxuu/fXp8kvP7La0X9Ot3TPKBvu0H+y2xVbNjjPR5aJIT+/szST7c93V+kv369W/qxzgnyTX9+zj7/F/rx7gkyd8v1o+2cFXlzduSN7p5Tv6Kbtrjf1+k3anAi/v7LwU+1t8/Cnj3As9ZB/xWf/8dwKXAfYAZusnGAPYHPjHynHcDR408f4+Rx3bvf25DNxfPYxdotw7YA/gZ4DJgZ2AX4Aq6mUxXAncD+/btTwGO7O9fD2zf399tntd0FHADcH9gR7rpAVYBP9m/R9v17f4G+LX+fgGHLfAe/dh4wFtH6tmN7vobOwO/Cxzfr39s/xpW9cu3j/R5KHBif/8fgKf291fQTdsB8Cbgs8D2/Xv1jf7fwE8BV8++nyPv+bz9eNuyb1O5eL22So8HLgYeDSx24ZcnA7/Y3/97ui2BcczOA3UZsEt11xe4Lcn3NmI//mHppu3dlm5KhsfQhctCngp8tKruAEjyEeBpfU1fqaqL+3YX0IUDfX/vT/Ix4GML9HtWVX1jpM+n0n0o/wxwfhLoQmJ93/4HdBPozWe+8Z5NNxnba/rlHeg+fJ8OvAugqi7tt+aWcgDwmL4mgF3Tz3VDN73BncCdSdbTTX/wDOBD1V80papuWayf/vepLZRBoEX1u3VOpJu99Wa6fexJNx/+k6vqu0t0Me4cJnf2P384cn92eVu6D9DRXZk7LFDvw4DXAE+sqlv7XR/zth192hh1QfdBPbtb62C6D9znA3+U5Keqm99q1NzXXv1Y76uq184z1veq6gcL1PFj4/V9/VJVXX2PF9N9CC/0vo+uH31f7sU8v8++r7nvwbb92PONMW8/2rJ5jECLqqqLq5v7/ot0f1n/C/Ccqtp3gf/sn6WbIRLgCLpLKW4O/0X3l+b2Se5Ld/B51m10u5IAdgXuAL6V5IF0c7nP127UZ4AXJtkp3UyPv8AiB8OT3AvYu6o+TXexlN3odinN9awku/fHFV4I/DtwNnBokgf0fe2e5KGLvfBFxjsDeGX6T+skjx95PUf0636abvfQrBuT/GTf5y+MrD+TbpLG2TH3Xaym/nUcluT+s69jI/vRFsAtAi0pyQxwa1X9MMmja/FrAh8NHJ/k94GbgJdsjhqq6tokp9DtIvkScNHIw2uA05PcUFU/n+Qiuv3819B9+M7bbqTvC/sth8/3q95TVRelu+rZfLYBTuoDKcA7av6zj86l2z32SOAfqmotQJLX01156l7AXcAr6IJuIfOOl+RP6K7cdWkfBuuAQ4C/BU7odwldPPK6AI4FPkF39b7L+VGAHQ38df+cbenC5DcXKqiqrkjyp8C/JvkB3e/jqA3tR1sGZx+VJiDJUXQHaH9nqbYD1HIO8JrZIJLmcteQJDXOLQJJapxbBJLUOINAkhpnEEhS4wwCSWqcQSBJjTMIJKlx/x9G0unxxewlAgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x432 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "classify_mutations(list_of_SNPS)\n",
    "plot_mutation_distribution(list_of_SNPS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[['V', 'R'], ['A', 'W']],\n",
       " [[], []],\n",
       " [['S'], ['S']],\n",
       " [[], []],\n",
       " [['A', 'W'], ['V', '*']],\n",
       " [['A'], ['P']],\n",
       " [['E', 'V', 'W'], ['E', 'V', 'R']],\n",
       " [['Y', 'L'], ['Y', 'L']],\n",
       " [['T', 'L', 'T'], ['T', 'P', 'T']],\n",
       " [['S'], ['N']],\n",
       " [['F', 'A', 'P', 'T'], ['S', 'T', 'P', 'M']],\n",
       " [['V'], ['V']],\n",
       " [[], []],\n",
       " [['G', 'S'], ['S', 'W']],\n",
       " [['S', 'W', 'T', 'L', 'V'], ['*', '*', 'T', '*', 'M']],\n",
       " [['A', 'V'], ['T', 'V']],\n",
       " [['T'], ['T']],\n",
       " [['P'], ['L']],\n",
       " [['S'], ['S']],\n",
       " [['T', 'L'], ['M', 'P']],\n",
       " [[], []],\n",
       " [[], []],\n",
       " [['W'], ['R']],\n",
       " [['T', 'L'], ['I', 'Q']],\n",
       " [['S', 'L', 'G', 'H', 'A'], ['S', 'V', 'V', 'N', 'A']],\n",
       " [[], []],\n",
       " [['R', 'N', 'R'], ['R', 'S', 'S']],\n",
       " [['A', 'R'], ['T', 'H']],\n",
       " [['P', 'P', 'H'], ['Q', 'L', 'R']],\n",
       " [['A'], ['A']],\n",
       " [['T', 'F'], ['T', 'L']],\n",
       " [['A', 'G'], ['T', 'G']],\n",
       " [[], []],\n",
       " [['W'], ['L']],\n",
       " [['R'], ['L']],\n",
       " [['I'], ['T']],\n",
       " [['T', 'P'], ['T', 'L']],\n",
       " [['M'], ['L']],\n",
       " [['S', 'V', 'H'], ['S', 'M', 'H']],\n",
       " [['P', 'P', 'V'], ['T', 'P', 'V']],\n",
       " [['A', 'D'], ['A', 'D']],\n",
       " [['R'], ['H']]]"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "amino_acid_mutations = find_AA_mutations_from_SNPs(list_of_SNPS,reference_gene)\n",
    "amino_acid_mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count_amino_acid_transitions(amino_acid_mutations, '*')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A 0.12092283214001591\n",
      "C 0.3269689737470167\n",
      "G 0.40731901352426414\n",
      "T 0.14478918058870327\n"
     ]
    }
   ],
   "source": [
    "for i in ['A', 'C', 'G', 'T']:\n",
    "    print(i, reference_gene.count(i)/len(reference_gene))"
   ]
  },{
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
